### PR TITLE
In-place receive for unstructured

### PR DESCRIPTION
**Pull request by bettiolm**

### CODE REVIEW

File: include/ghex/communication_object_ipr.hpp
Comment (mbianco) [inline include/ghex/communication_object_ipr.hpp:14]: Check this before merging

File: include/ghex/communication_object_ipr.hpp
Comment (mbianco) [inline include/ghex/communication_object_ipr.hpp:32]: remove? Ho do we protect against mis-use?

File: include/ghex/communication_object_ipr.hpp
Comment (mbianco) [inline include/ghex/communication_object_ipr.hpp:130]: identation

File: tests/unstructured_user_concepts.cpp
Comment (mbianco) [inline tests/unstructured_user_concepts.cpp:525]:
```suggestion
```

**Review by mbianco — COMMENTED** [review]
I'm very slow on this one. Some  minor comments for now

File: include/ghex/communication_object_ipr.hpp
Comment (mbianco) [inline include/ghex/communication_object_ipr.hpp:145]: are these ones used for unpack? I guess not.

File: include/ghex/communication_object_ipr.hpp
Comment (mbianco) [inline include/ghex/communication_object_ipr.hpp:150]: Also this struct is the same as the co_2. Would it make sense to use some composition to reuse some code? (how much code would be saved by doing so?)

File: include/ghex/communication_object_ipr.hpp
Comment (mbianco) [inline include/ghex/communication_object_ipr.hpp:161]: This struct also is the same as co_2. Maybe instead of composing classes, we can put these sub-structs into an impl namespace in a separate file and use them when we need them

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (boeschf) [inline include/ghex/unstructured/communication_object_ipr.hpp:53]: the friend is needed because of private ctor

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (boeschf) [inline include/ghex/unstructured/communication_object_ipr.hpp:57]: comm is probably not needed

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (boeschf) [inline include/ghex/unstructured/communication_object_ipr.hpp:382]: is this resolved?

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/unstructured/communication_object_ipr.hpp:51]: Right. Done.

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/unstructured/communication_object_ipr.hpp]: Removed from class members.

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/unstructured/communication_object_ipr.hpp:372]: Yes, I just left it behind

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [inline include/ghex/unstructured/communication_object_ipr.hpp:160]: Is it ok to stick with this naming convention (same used in `communication_object_2`)?

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [inline include/ghex/unstructured/communication_object_ipr.hpp:198]: Same comment here on the naming convention

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [inline include/ghex/unstructured/communication_object_ipr.hpp:322]: Here I used the same constructor as in `communication_object_2`, i.e. the one with `handle` as argument, which moves only the handle. Not a big deal, but I think that using directly the future move constructor could also be an option, since we are using the same future specialization in both cases anyway. Probably this could be a bit too strict if one day we should decide (for some reason) not to use the `future<void>` specialization as return type for the send / recv functions.

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (mbianco) [reply include/ghex/unstructured/communication_object_ipr.hpp:160]: I did not notice the names here. We use `m_` for members that are not assumed to be accessed publicly.
Anyway, since these members are not accessed in this PR, I would leave the code as is, since otherwise you have to make more diffs.

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (mbianco) [reply include/ghex/unstructured/communication_object_ipr.hpp:198]: as above

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (mbianco) [reply include/ghex/unstructured/communication_object_ipr.hpp:322]: I'm not sure I understand completely. The trade-off between reuse of patterns and code is tricky. I would not spend too much thinking about possible future choices, and maybe put a comment explaining what to do if an error appears about it

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/unstructured/communication_object_ipr.hpp:160]: Ok, sure

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/unstructured/communication_object_ipr.hpp:198]: Yep

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/unstructured/communication_object_ipr.hpp]: Ok, I just added a comment in the code

File: include/ghex/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/communication_object_ipr.hpp]: Done, added a couple of issues as well.

File: include/ghex/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/communication_object_ipr.hpp]: Done, used a specialization for unstructured.

File: include/ghex/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/communication_object_ipr.hpp]: Done

File: tests/unstructured_user_concepts.cpp
Comment (bettiolm) [reply tests/unstructured_user_concepts.cpp]: Done

File: include/ghex/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/communication_object_ipr.hpp]: This struct is actually used for the receive part as well, using `std::function<void()>` as function type. I first thought of not using a func at all, but this requires at least to define a new struct (or a specialization of this one), and I think that keeping it as it is is less confusing in the end. I'll mention this in one of the two opened issues anyway.

File: include/ghex/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/communication_object_ipr.hpp]: Moved to refactoring issue.

File: include/ghex/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/communication_object_ipr.hpp]: Moved to refactoring issue.

File: include/ghex/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/communication_object_ipr.hpp]: I changed the code a little bit here now. `field_info` is actually used only for the send part, so only with pack function. I added a specific struct for the receive, without callback (unpack) function.

**Comment by bettiolm** [comment]
I addressed the comments and added a couple of issues (mostly minor refactoring) which can be addresses after the merge in my opinion. From my side it is ok now.

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (mbianco) [inline include/ghex/unstructured/communication_object_ipr.hpp:389]: why is this comment useful? could you expand it?

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (mbianco) [inline include/ghex/unstructured/communication_object_ipr.hpp:423]: why is this comment useful? could you expand it?

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (mbianco) [inline include/ghex/unstructured/communication_object_ipr.hpp:483]: why is this comment useful? could you expand it?

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/unstructured/communication_object_ipr.hpp:389]: It's just a reminder, memory is already a pointer, and it is used as such inside, so a copy could also be ok. Same happens in comm2. I just did not want to change much the interface with respect of the comm2 one, since that should be harmless.

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/unstructured/communication_object_ipr.hpp:423]: as before

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/unstructured/communication_object_ipr.hpp:483]: as before

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/unstructured/communication_object_ipr.hpp]: I pushed some documentation and removed the pass by reference from this function (allocate), in which it is actually not needed.

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/unstructured/communication_object_ipr.hpp]: Actually not, sorry. Here memory is not a pointer as before, and represents the recv memory. Reference is needed. I pushed the updated version.

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/unstructured/communication_object_ipr.hpp]: Same as for `set_recv_size`

File: include/ghex/unstructured/communication_object_ipr.hpp
Comment (bettiolm) [reply include/ghex/unstructured/communication_object_ipr.hpp]: From my side it should be ok now

**Review by mbianco — APPROVED** [review]

**Pull request merged**
